### ISSUE TITLE
Failure to consume event when writing data in Firestore emulator: "Invalid pattern. Reason: [85:-] Expected '}' at end of capture expression."

### ISSUE THREAD

**Issue by Nevvulo** [firebase/firebase-tools#3136]
### [REQUIRED] Environment info

**firebase-tools:** v9.3.0

**Platform:** Windows 10 Home, version 2004

### [REQUIRED] Test case
I started the Firestore emulator using `firebase emulators:start --export-on-exit --import ~/.flux` and everything loads as expected, no warnings. Navigating to the Emulator UI (which is hosted at `localhost:4000` for me), if I modify/create a field or document (basically mutate data in any way) - it will process the first event to mutate the data but any subsequent events will be ignored. In my testing I have noticed this is not exclusive to the Emulator UI, as you'll see below there are errors in the `firestore-debug.log` file when trying to process events.

### [REQUIRED] Steps to reproduce
1. Launch the emulator using `firebase emulators:start --export-on-exit --import [directory]`
2. Create a document (ID doesn't matter) with a field (name/value doesn't matter)
3. Inside the same document, create another field (name/value doesn't matter)
4. Refresh the Emulator UI window and you should see that only the first field was saved to the document

### [REQUIRED] Expected behavior
I would expect that all data that is written and valid is saved in the local Firestore emulator.

### [REQUIRED] Actual behavior
Any data that is written after the first write in a session is not processed and therefore not saved to the database.
![2021-02-15_17-14-12](https://user-images.githubusercontent.com/26913683/107913002-bb788900-6fb3-11eb-9064-ca43dd8952cb.gif)

I am consistently receiving this error in my `firestore-debug.log` and when running `firestore emulators:start` with the `--debug` flag:
```
WARNING: Failed to consume the event.
java.lang.IllegalArgumentException: Invalid pattern. Reason: [85:-] Expected '}' at end of capture expression.
[85:-ww}] Unrecognized character(s) at end of path.
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:142)
        at com.google.firebase.rules.eventflow.client.path.PathPattern.compileInternal(PathPattern.java:78)
        at com.google.firebase.rules.eventflow.client.path.PathPattern.compile(PathPattern.java:50)
        at com.google.firebase.rules.eventflow.client.EventRuleEvaluatorImpl.lambda$new$0(EventRuleEvaluatorImpl.java:58)
        at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:412)
        at com.google.firebase.rules.eventflow.client.EventRuleEvaluatorImpl.<init>(EventRuleEvaluatorImpl.java:52)
        at com.google.firebase.rules.eventflow.client.EventRuleEvaluatorImpl.<init>(EventRuleEvaluatorImpl.java:39)
        at com.google.cloud.datastore.emulator.impl.events.EventManager.reportEvents(EventManager.java:84)
        at com.google.cloud.datastore.emulator.impl.CloudFirestoreV1.commitHelper(CloudFirestoreV1.java:835)
        at com.google.cloud.datastore.emulator.impl.CloudFirestoreV1.internalCommit(CloudFirestoreV1.java:731)
        at com.google.cloud.datastore.emulator.impl.CloudFirestoreV1.internalCommit(CloudFirestoreV1.java:715)
        at com.google.cloud.datastore.emulator.impl.CloudFirestoreV1.write(CloudFirestoreV1.java:704)
        at com.google.cloud.datastore.emulator.impl.CloudFirestoreV1WriteStream.handleRequest(CloudFirestoreV1WriteStream.java:208)
        at com.google.cloud.datastore.emulator.impl.CloudFirestoreV1WriteStream.write(CloudFirestoreV1WriteStream.java:136)
        at com.google.cloud.datastore.emulator.impl.CloudFirestoreV1WriteStream.onNext(CloudFirestoreV1WriteStream.java:92)
        at com.google.cloud.datastore.emulator.impl.CloudFirestoreV1WriteStream.onNext(CloudFirestoreV1WriteStream.java:25)
        at com.google.cloud.datastore.emulator.firestore.webchannel.FirestoreV1WebChannelAdapter$FirestoreWriteHandler.onMessage(FirestoreV1WebChannelAdapter.java:141)
        at com.google.cloud.datastore.emulator.firestore.webchannel.FirestoreV1WebChannelAdapter$FirestoreWriteHandler.onMessage(FirestoreV1WebChannelAdapter.java:111)
        at com.google.net.webchannel.server.v8.ChannelInternalImpl.onMessage(ChannelInternalImpl.java:485)
        at com.google.net.webchannel.server.v8.ForwardChannel.deliverMessage(ForwardChannel.java:114)
        at com.google.net.webchannel.server.v8.ForwardChannel.processPendingRequest(ForwardChannel.java:93)
        at com.google.net.webchannel.server.v8.ForwardChannel.consume(ForwardChannel.java:67)
        at com.google.net.webchannel.server.v8.ForwardChannel.add(ForwardChannel.java:60)
        at com.google.net.webchannel.server.v8.ChannelInternalImpl.handleRequestParams(ChannelInternalImpl.java:260)
        at com.google.net.webchannel.server.v8.ChannelInternalImpl.processSessionRequest(ChannelInternalImpl.java:317)
        at com.google.net.webchannel.server.v8.ChannelInternalImpl.handle(ChannelInternalImpl.java:142)
        at com.google.net.webchannel.server.v8.ChannelInternalImpl.consume(ChannelInternalImpl.java:78)
        at com.google.net.webchannel.server.common.ChannelProcessor.run(ChannelProcessor.java:80)
```
I have not changed my local environment much since this issue started occuring (today), and the stack trace appears to be complaining about something wrong with `firestore.rules` but I haven't touched this file in a while and I can confirm the file is syntactically valid. I am also able to reproduce this on another Windows 10 Pro machine, version 2004 in the same project folder. Please let me know if you need any more logs or information, thanks 🙏🏼

**Comment by samtstern** [comment]
@Nevvulo sounds like you have a malformed cloud functions trigger or Firestore rule that is missing a `}` character. Can you show me your rules file and also any Cloud Functions definitions involving Firestore>

**Comment by Nevvulo** [comment]
@samtstern Thanks for the quick response, you're absolutely right. Turns out it wasn't as black and white as that, I had a cloud function trigger that looked like this `data/{uid}/{yyyy-ww}` and it was the dash character that was causing the pattern to mess up, hence the `[85:-ww}]` in the error. 

Appreciate your help 🤗 however if I had to give some feedback it would be nice if there was a less verbose error especially since I wasn't aware you couldn't have dashes in capture expressions. Regardless I'll close this issue as this has been resolved, thanks again!
